### PR TITLE
Add article: Why I enjoy working with ReactJS

### PR DESCRIPTION
**Pull request by kjcpaas**
Article about understanding React components, and why I love working with ReactJS.

Enjoy!

### CODE REVIEW

**Comment by kjcpaas** [comment]
@quipper/devs Please review 😸

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+```
+
+For those who are new to ReactJS, seeing the markup in the same file as the logic might be weird at first since it's usually found in a separate HTML file (or another file type, depending on template engine used).
+
+However, the biggest benefit of having the markup with the same file as the logic is that there is no need to navigate a project's file tree (or the DOM in the browser) just to look for the HTML element that corresponds to a piece of code. The time used for file navigation can be instead used for writing code. This contributes a lot to efficiency.
+
+
+## How it internally works
+
+Abstraction is very important in software engineering, in general. However, there are times when developers need to dive deeper in to the code to investogate a bug, or enhance a feature. Hence, developers need to easily understand how a component works internally.
```
Comment (banyan) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: `s/investogate/investigate/`

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+`props` are properties passed by the parent component, while the `state` are properties that are only accessed by the component itself.
+
+Default values of `props` can be set by `defaultProps`, while the types can be restricted by `propTypes`. Using `propTypes` adds validation to the component to make sure the data being passed is compatible with the component.
+
+`state` cannot be assigned directly, except inside `constructor` method where it is initialized. To update the state `this.setState` must be used.
+
+### Lifecycle
+
+The are 3 major steps in React's component lifecycle: **mounting**, **updating**, and **unmounting**. It's straightforward and easy to understand because of the use of `will` and `did` in method names for methods that gets executed around `render`.
```
Comment (banyan) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: `s/The/There/`

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+And yes, `ToggleButton` is used just like how an HTML tag is used!
+
+Seeing components used like this makes it look neat. It's also easy to see where a component and logic goes in your DOM.
+
+
+## Conclusion
+
+ReactJS, for me, is built for builders like myself. The component is well designed because of easy integration with other components. We can also easily 'tinker', if needed be, since we can easily know a lot (if not all) of information about that component by just looking at its file.
+
+It reminds me of the Lego blocks I used to play when I was a child. After all, maybe the biggest reason I enjoy working with ReactJS is that it reminds me of the fun I had as a child.
```
Comment (banyan) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: great wrapup ❤️

**Comment by banyan** [comment]
LGTM 👍

**Comment by kjcpaas** [comment]
@banyan Typos fixed. Please check again cc @quipper/devs 😸

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+layout: post
+title:  "Why I enjoy working with ReactJS"
+date:   2016-11-25 10:00:00
+author: kjcpaas
+comments: true
+---
+
+ReactJS by Facebook is one of the most popular Javascript libraries nowadays.
+
+Many developers love it for many reasons, from efficiency to even its tools. I personally like because of how components are built in `jsx`. In one glance, one can easily see 3 very important things: how it looks like (in the DOM), how it internally works, and how it can connect to other components. As someone who loves building things and investigating how something works, this is a very, very important point for me.
```
Comment (bdesmero) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: `I personally like it` 😁

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+`props` are properties passed by the parent component, while the `state` are properties that are only accessed by the component itself.
+
+Default values of `props` can be set by `defaultProps`, while the types can be restricted by `propTypes`. Using `propTypes` adds validation to the component to make sure the data being passed is compatible with the component.
+
+`state` cannot be assigned directly, except inside `constructor` method where it is initialized. To update the state `this.setState` must be used.
+
+### Lifecycle
+
+There are 3 major steps in React's component lifecycle: **mounting**, **updating**, and **unmounting**. It's straightforward and easy to understand because of the use of `will` and `did` in method names for methods that gets executed around `render`.
```
Comment (bdesmero) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: `methods that get executed`? 😁

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown (hunk 3 above)
Comment (bdesmero) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: Indeed! 🎉

**Comment by bdesmero** [comment]
Good read! Thanks for sharing!!! 💃

File: _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+date:   2016-11-25 10:00:00
+author: kjcpaas
+comments: true
+---
+
+ReactJS by Facebook is one of the most popular Javascript libraries nowadays.
+
+Many developers love it for many reasons, from efficiency to even its tools. I personally like because of how components are built in `jsx`. In one glance, one can easily see 3 very important things: how it looks like (in the DOM), how it internally works, and how it can connect to other components. As someone who loves building things and investigating how something works, this is a very, very important point for me.
+
+Let's consider a toggle button. It's React component code will look like below:
```
Comment (sethjeffery) [inline _posts/2016-11-25-why-i-enjoy-working-with-react-js.markdown]: Its

**Comment by sethjeffery** [comment]
I'm no JSXpert but I noticed in all your code examples, there were no semicolons. The examples I've seen online have shown semicolon use. Is it optional? Is it just code style?

**Comment by kjcpaas** [comment]
@sethjeffery Even in js, semicolons are optional but after ES6 release, more and more people do not use semicolons.

In Qlink-react we just decided to not use it.

**Comment by banyan** [comment]
Just examples of other cases.

* https://github.com/ReactTraining/react-router#whats-it-look-like
* https://github.com/reactjs/redux#the-gist

**Comment by kjcpaas** [comment]
@bdesmero @sethjeffery Fixed the grammar errors you pointed out. Thanks a lot!

**Comment by banyan** [comment]
:+1:

**Comment by kjcpaas** [comment]
Thanks!

**Pull request merged**